File: src/main/java/com/kritsn/leetCodeJava/_637AverageOfLevelsInBinaryTree.java
package com.kritsn.leetCodeJava;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * Copyright © 2026 Kritsn LLP. All rights reserved.
 *
 * @author Radhey (hr-sh)
 * @since Jul 20, 2026
 */

/*
Leetcode 637: Average of Levels in Binary Tree

Given the root of a binary tree, return the average value of the nodes on each level in the form of an array.
Answers within 10^-5 of the actual answer will be accepted.
*/
public class _637AverageOfLevelsInBinaryTree {

    // ✅ Definition for a binary tree node
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }
    }

    /**
     * 🧠 Algorithm & Approach:
     * - Perform a level order traversal using a queue.
     * - At each level, sum the values of all nodes.
     * - Compute average by dividing sum by the number of nodes at that level.
     * - Append average to result list.
     * <p>
     * Time Complexity: O(n) — visiting every node once.
     * Space Complexity: O(n) — for queue and result storage.
     */
    List<Double> averageOfLevels(TreeNode root) {
        List<Double> result = new ArrayList<>();
        if (root == null) return result;

        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            double levelSum = 0.0;

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.removeFirst();
                levelSum += node.val;

                if (node.left != null) queue.add(node.left);
                if (node.right != null) queue.add(node.right);
            }

            result.add(levelSum / levelSize);
        }

        return result;
    }

    public static void main(String[] args) {
        _637AverageOfLevelsInBinaryTree sol = new _637AverageOfLevelsInBinaryTree();

        // 🧪 Test Case 1
        TreeNode root1 = new TreeNode(3);
        root1.left = new TreeNode(9);
        root1.right = new TreeNode(20);
        root1.right.left = new TreeNode(15);
        root1.right.right = new TreeNode(7);
        System.out.println("Test Case 1: " + sol.averageOfLevels(root1)); // Expected: [3.0, 14.5, 11.0]

        // 🧪 Test Case 2
        TreeNode root2 = new TreeNode(10);
        System.out.println("Test Case 2: " + sol.averageOfLevels(root2)); // Expected: [10.0]

        // 🧪 Test Case 3
        System.out.println("Test Case 3: " + sol.averageOfLevels(null)); // Expected: []
    }
}

File: src/main/java/com/kritsn/leetcode/_637AverageOfLevelsInBinaryTree.kt
package com.kritsn.leetcode

///////////////////////////////////////////////////////////////////////////
//           Copyright © 2025 Kritsn LLP. All rights reserved.
//                      @author Radhey (hr-sh)
//                      @since Jul 24, 2025
///////////////////////////////////////////////////////////////////////////
/*
Leetcode 637: Average of Levels in Binary Tree

Given the root of a binary tree, return the average value of the nodes on each level in the form of an array.
Answers within 10^-5 of the actual answer will be accepted.
*/

class _637AverageOfLevelsInBinaryTree {

    // ✅ Definition for a binary tree node
    data class TreeNode(var `val`: Int, var left: TreeNode? = null, var right: TreeNode? = null)

    /**
     * 🧠 Algorithm & Approach:
     * - Perform a level order traversal using a queue.
     * - At each level, sum the values of all nodes.
     * - Compute average by dividing sum by the number of nodes at that level.
     * - Append average to result list.
     *
     * Time Complexity: O(n) — visiting every node once.
     * Space Complexity: O(n) — for queue and result storage.
     */
    fun averageOfLevels(root: TreeNode?): List<Double> {
        val result = mutableListOf<Double>()
        if (root == null) return result

        val queue = ArrayDeque<TreeNode>()
        queue.add(root)

        while (queue.isNotEmpty()) {
            val levelSize = queue.size
            var levelSum = 0.0

            repeat(levelSize) {
                val node = queue.removeFirst()
                levelSum += node.`val`

                node.left?.let { queue.add(it) }
                node.right?.let { queue.add(it) }
            }

            result.add(levelSum / levelSize)
        }

        return result
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val sol = _637AverageOfLevelsInBinaryTree()

            // 🧪 Test Case 1
            val root1 = TreeNode(3).apply {
                left = TreeNode(9)
                right = TreeNode(20).apply {
                    left = TreeNode(15)
                    right = TreeNode(7)
                }
            }
            println("Test Case 1: ${sol.averageOfLevels(root1)}") // Expected: [3.0, 14.5, 11.0]

            // 🧪 Test Case 2
            val root2 = TreeNode(10)
            println("Test Case 2: ${sol.averageOfLevels(root2)}") // Expected: [10.0]

            // 🧪 Test Case 3
            println("Test Case 3: ${sol.averageOfLevels(null)}") // Expected: []
        }
    }
}